File: src/features/workouts/__tests__/payload.test.ts
import type { Exercise, GateDetail, GateTemplate, TemplateExercise } from '@/features/gates/types';
import {
  completeSet,
  createActiveWorkout,
  toggleWarmup,
  updateSet,
} from '@/features/workouts/logic';
import { buildCompletionPayload } from '@/features/workouts/payload';

function detail(): GateDetail {
  const template: GateTemplate = {
    id: 'tpl1',
    user_id: null,
    name: 'Push',
    description: 'Chest',
    estimated_duration_minutes: 50,
    difficulty: 'C',
    is_system_template: true,
    created_at: '',
    updated_at: '',
  };
  const mkEx = (exId: string, name: string): TemplateExercise & { exercise: Exercise } => ({
    id: `te_${exId}`,
    template_id: 'tpl1',
    exercise_id: exId,
    order_index: 0,
    target_sets: 2,
    target_reps_min: 5,
    target_reps_max: 8,
    target_rpe: 8,
    rest_seconds: 90,
    exercise: {
      id: exId,
      name,
      category: 'chest',
      primary_muscle_group: 'chest',
      secondary_muscle_groups: [],
      equipment: 'barbell',
      exercise_type: 'strength',
      ranking_enabled: true,
      created_at: '',
    },
  });
  return { template, exercises: [mkEx('ex1', 'Bench'), mkEx('ex2', 'OHP')] };
}

describe('buildCompletionPayload', () => {
  it('persists only completed sets, only exercises with any, and computes aggregates', () => {
    let w = createActiveWorkout(detail(), 0); // startedAt = epoch 0
    // ex1: a warm-up set then a working set, both completed
    w = updateSet(w, 0, 0, { weightKg: 60, reps: 5 });
    w = toggleWarmup(w, 0, 0);
    w = completeSet(w, 0, 0, 1000);
    w = updateSet(w, 0, 1, { weightKg: 100, reps: 5 });
    w = completeSet(w, 0, 1, 2000);
    // ex2: nothing completed

    const { payload, aggregates } = buildCompletionPayload(w, 60_000);

    expect(payload.exercises).toHaveLength(1); // ex2 omitted
    expect(payload.exercises[0].sets).toHaveLength(2);
    expect(payload.exercises[0].sets.map((s) => s.set_number)).toEqual([1, 2]);

    // Volume + count exclude the warm-up.
    expect(aggregates.totalVolumeKg).toBe(500); // 100 * 5
    expect(aggregates.completedSets).toBe(1);
    expect(aggregates.exerciseCount).toBe(1);
    expect(aggregates.durationSeconds).toBe(60);

    // est-1RM: working set computed, warm-up left null.
    const working = payload.exercises[0].sets.find((s) => !s.is_warmup)!;
    const warm = payload.exercises[0].sets.find((s) => s.is_warmup)!;
    expect(working.estimated_1rm_kg).toBe(116.67);
    expect(warm.estimated_1rm_kg).toBeNull();

    // Session carries the idempotency id + completed status fields.
    expect(payload.session.id).toBe(w.sessionId);
    expect(payload.session.total_volume_kg).toBe(500);
  });

  it('produces no exercises when nothing was completed', () => {
    const w = createActiveWorkout(detail(), 0);
    const { payload, aggregates } = buildCompletionPayload(w, 1000);
    expect(payload.exercises).toHaveLength(0);
    expect(aggregates.completedSets).toBe(0);
  });
});

File: android-native/app/src/test/java/com/fitnessrpg/app/domain/workouts/PayloadTest.kt
package com.fitnessrpg.app.domain.workouts

import com.fitnessrpg.app.domain.model.Exercise
import com.fitnessrpg.app.domain.model.GateDetail
import com.fitnessrpg.app.domain.model.GateTemplate
import com.fitnessrpg.app.domain.model.TemplateExercise
import com.fitnessrpg.app.domain.model.TemplateExerciseWithExercise
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class PayloadTest {

    private fun mkEx(exId: String, name: String) = TemplateExerciseWithExercise(
        templateExercise = TemplateExercise("te_$exId", "tpl1", exId, 0, 2, 5, 8, 8.0, 90),
        exercise = Exercise(exId, name, "chest", "chest", emptyList(), "barbell", "strength", true, ""),
    )

    private fun detail() = GateDetail(
        template = GateTemplate("tpl1", null, "Push", "Chest", 50, "C", true, "", ""),
        exercises = listOf(mkEx("ex1", "Bench"), mkEx("ex2", "OHP")),
    )

    @Test
    fun `persists only completed sets, only exercises with any, and computes aggregates`() {
        var w = createActiveWorkout(detail(), 0L) // startedAt = epoch 0
        w = updateSet(w, 0, 0, SetPatch(weightKg = 60.0, reps = 5))
        w = toggleWarmup(w, 0, 0)
        w = completeSet(w, 0, 0, 1000L)
        w = updateSet(w, 0, 1, SetPatch(weightKg = 100.0, reps = 5))
        w = completeSet(w, 0, 1, 2000L)
        // ex2: nothing completed

        val (payload, aggregates) = buildCompletionPayload(w, 60_000L)

        assertEquals(1, payload.exercises.size) // ex2 omitted
        assertEquals(2, payload.exercises[0].sets.size)
        assertEquals(listOf(1, 2), payload.exercises[0].sets.map { it.setNumber })

        assertEquals(500.0, aggregates.totalVolumeKg, 1e-9) // 100 * 5
        assertEquals(1, aggregates.completedSets)
        assertEquals(1, aggregates.exerciseCount)
        assertEquals(60, aggregates.durationSeconds)

        val working = payload.exercises[0].sets.first { !it.isWarmup }
        val warm = payload.exercises[0].sets.first { it.isWarmup }
        assertEquals(116.67, working.estimated1rmKg!!, 1e-9)
        assertNull(warm.estimated1rmKg)

        assertEquals(w.sessionId, payload.session.id)
        assertEquals(500.0, payload.session.totalVolumeKg, 1e-9)
    }

    @Test
    fun `produces no exercises when nothing was completed`() {
        val w = createActiveWorkout(detail(), 0L)
        val (payload, aggregates) = buildCompletionPayload(w, 1000L)
        assertEquals(0, payload.exercises.size)
        assertEquals(0, aggregates.completedSets)
    }
}
